burgerSync.js

```diff
@@ -0,0 +1,252 @@
+    initializeConnection(peer) {
+        this._printMessage('Peer Connected: ' + this._getInternalQualifiedName(peer));
+        this._write(peer, this.MESSAGE_TYPE.INITIAL_HANDSHAKE_QUERY, 'Hello!');
+        peer.on('close', () => {
+            this.initializeErrorHandler(peer);
+        });
+        peer.on('error', () => {
+            this.initializeErrorHandler(peer);
+        });
+        peer.on('message', (message) => {
+            this.initializeMessageHandler(peer, message);
+        });
+    }
+
+    /**
+     * Handles the messages received from the websockets.
+     * For the uniformity of the messages, this schema
+     * is followed:
+     * 
+     * schema = {
+     *      type: this.MESSAGE_TYPE[the-message-type],
+     *      message: the message object
+     * }
+     * 
+     * Usage of burgerSync._write() method is recommended
+     * to send message.
+     * 
+     * @param {websocket} peer - the socket connection object to the peer
+     * @param {object} rawMessage - the message received on the socket
+     */
+    initializeMessageHandler(peer, rawMessage) {
+        let transmission = JSON.parse(rawMessage);
+        let message = transmission.message;
+
+        switch(transmission.type) {
+            case this.MESSAGE_TYPE.INITIAL_HANDSHAKE_QUERY: 
+                this._printMessage('Initial handshake Sent!')
+                this._write(peer, this.MESSAGE_TYPE.INITIAL_HANDSHAKE_RESPONSE, this.burgerNode.info);
+                break; 
+            case this.MESSAGE_TYPE.INITIAL_HANDSHAKE_RESPONSE: 
+                this._printMessage('Initial handshake received!');
+                if (message.cumulativeDifficulty > this.burgerNode.info.cumulativeDifficulty) {
+                    this._printMessage('Peer chain is better, requesting sync...');
+                    this._write(peer, this.MESSAGE_TYPE.REQUEST_SYNC_CHAIN, 'Let us sync!');
+                }
+                break;
+            case this.MESSAGE_TYPE.REQUEST_SYNC_CHAIN:
+                this._printMessage('Sync request received! Sending chain...')
+                this._write(peer, this.MESSAGE_TYPE.RESPONSE_SYNC_CHAIN, this.burgerNode.chain);
+                break;
+            case this.MESSAGE_TYPE.RESPONSE_SYNC_CHAIN:
+                /**
+                 * When chain is synced, all pending transactions are lost.
+                 * Wait for Issue #41 and adjust needed logic to concat
+                 * pending transactions instead of replacing.
+                 */
+                this._printMessage('Chain received! Syncing...')
+                this.burgerNode.replaceChain(message);
+                this._printMessage('    Synced!');
+                break;
+            case this.MESSAGE_TYPE.BROADCAST_NEW_BLOCK:
+                this._printMessage('A new block has been mined!');
+                if (message.cumulativeDifficulty > this.burgerNode.info.cumulativeDifficulty) {
+                    this.burgerNode.replaceChain(message);
```

Comment:
@24thsaint I'm worried that this is replacing the entire chain than just potentially appending the new block to our otherwise valid chain. Thoughts?

Reply:
Yeah. I was worried about this, too. The closest one we've got for the append functionality is `BurgerBlockchain.addBlock()` but it still lacks validations so I just used the `replaceChain()`.

Reply:
Maybe we will work that in eventually? I'd like there to be some sort of append block mechanism in some form.